src/casync-http.c

```diff
@@ -11,67 +12,1066 @@
+};
+
+typedef struct Queue {
+        QueueItem *head;
+        QueueItem *tail;
+        uint64_t len;
+        uint64_t n_added;   /* total number of items added */
+        uint64_t n_removed; /* total number of items removed */
+} Queue;
+
+static int queue_push(Queue *q, void *data) {
+        int r;
+        QueueItem *item;
+
+        assert(q);
+        assert(data);
+
+        item = new0(QueueItem, 1);
+        if (!item) {
+                r = log_oom();
+                return r;
+        }
+
+        item->data = data;
+        item->next = NULL;
+
+        if (q->tail) {
+                q->tail->next = item;
+                q->tail = item;
+        } else {
+                assert(!q->head);
+                q->head = item;
+                q->tail = item;
+        }
+
+        q->n_added++;
+        q->len++;
+
+        return 0;
+}
+
+static void *queue_pop(Queue *q) {
+        QueueItem *item;
+        void *data;
+
+        assert(q);
+
+        item = q->head;
+        if (!item)
+                return NULL;
+
+        q->head = item->next;
+        if (!q->head)
+                q->tail = NULL;
+
+        data = item->data;
+        free(item);
+
+        q->n_removed++;
+        q->len--;
+
+        return data;
+}
+
+static void *queue_remove(Queue *q, void *data) {
+        QueueItem *curr, *prev;
+
+        assert(q);
+
+        for (prev = NULL, curr = q->head; curr; prev = curr, curr = curr->next)
+                if (curr->data == data)
+                        break;
+
+        if (curr == NULL)
+                return NULL; /* not found */
+
+        if (prev == NULL)
+                q->head = curr->next; /* removing first item */
+        else
+                prev->next = curr->next; /* removing later item */
+
+        if (q->tail == curr)
+                q->tail = prev; /* removed last item */
+
+        free(curr);
+
+        q->n_removed++;
+        q->len--;
+
+        return data;
+}
+
+static uint64_t queue_len(Queue *q) {
+        assert(q);
+
+        return q->len;
+}
+
+static bool queue_is_empty(Queue *q) {
+        assert(q);
+
+        return q->head == NULL;
+}
+
+static void queue_free(Queue *q) {
+        if (q == NULL)
+                return;
+
+        free(q);
+}
+
+static Queue *queue_new(void) {
+        return new0(Queue, 1);
+}
+
+/*
+ * Chunk Downloader
+ *
+ * We re-use things as much as possible, which means that:
+ * - CURL handles are allocated once at the beginning, then re-used all along.
+ * - ChunkData objects (ie. ReallocBuffer) as well.
+ *
+ * During operations, our CURL handles move from one queue to another, ie:
+ *   ready -> inprogress -> completed -> ready ...
+ *
+ * TODO This queues could be static (ie. of fixed size) to avoid a bunch of malloc/free.
+ */
+
+typedef struct CaChunkDownloader CaChunkDownloader;
+
+struct CaChunkDownloader {
+        CaRemote *remote;
+        CURLM *multi;
+        Queue *ready;       /* CURL handles waiting to be used */
```

Comment:
Since we know the size of the queues already, it would be more efficient to have static arrays, and avoid a lot of malloc/free. I can rework this.